contracts/FeralfileArtworkV4_2.sol

```diff
@@ -0,0 +1,184 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.13;
+
+import {Nonces} from "./Nonces.sol";
+import {FeralfileExhibitionV4_1} from "./FeralfileArtworkV4_1.sol";
+import {IFeralfileVaultV2} from "./IFeralfileVaultV2.sol";
+import {FeralfileSaleDataV2} from "./FeralfileSaleDataV2.sol";
+
+contract FeralfileExhibitionV4_2 is
+    FeralfileExhibitionV4_1,
+    FeralfileSaleDataV2,
+    Nonces
+{
+    error TokenIDNotFound();
+    error FunctionNotSupported();
+
+    // vault contract instance
+    IFeralfileVaultV2 public immutable VAULT_V2;
+
+    mapping(uint256 => uint256) private seriesNextSaleTokenIds; // seriesID -> tokenID
+
+    constructor(
+        string memory name_,
+        string memory symbol_,
+        bool burnable_,
+        bool bridgeable_,
+        address signer_,
+        address vault_,
+        address costReceiver_,
+        string memory contractURI_,
+        uint256[] memory seriesIds_,
+        uint256[] memory seriesMaxSupplies_,
+        uint256[] memory seriesNextSaleTokenIds_
+    )
+        FeralfileExhibitionV4_1(
+            name_,
+            symbol_,
+            burnable_,
+            bridgeable_,
+            signer_,
+            vault_,
+            costReceiver_,
+            contractURI_,
+            seriesIds_,
+            seriesMaxSupplies_
+        )
+    {
+        VAULT_V2 = IFeralfileVaultV2(payable(vault_));
+        for (uint256 i = 0; i < seriesIds_.length; i++) {
+            seriesNextSaleTokenIds[seriesIds_[i]] = seriesNextSaleTokenIds_[i];
+        }
+    }
+
+    /// @notice pay to get artworks to a destination address. The pricing, costs and other details is included in the saleData
+    /// @param r_ - part of signature for validating parameters integrity
+    /// @param s_ - part of signature for validating parameters integrity
+    /// @param v_ - part of signature for validating parameters integrity
+    /// @param saleData_ - the sale data
+    function buyBulkArtworks(
+        bytes32 r_,
+        bytes32 s_,
+        uint8 v_,
+        SaleDataV2 calldata saleData_
+    ) external payable {
+        require(_selling, "FeralfileExhibitionV4: sale is not started");
+        super._checkContractOwnedToken();
+        validateSaleDataV2(saleData_);
+
+        saleData_.payByVaultContract
+            ? VAULT_V2.payForSaleV2(r_, s_, v_, saleData_)
+            : require(
+                saleData_.price == msg.value,
+                "FeralfileExhibitionV4: invalid payment amount"
+            );
+
+        bytes32 message = keccak256(
+            abi.encode(block.chainid, address(this), saleData_)
+        );
+
+        //check nonce
+        _useCheckedNonce(saleData_.destination, saleData_.nonce);
```

Comment:
It should be after the signature is validated. Otherwise, we could waste a little bit gas if the signature is invalid/expired.